### PR TITLE
fix(Attachments-UI):Restrict from duplicate attachment upload

### LINKED ISSUE #417: There should not be duplicate attachment names
Currently, a user could make an upload of attachments with the same name multiple times.

The system should / could enforce different names

### PR DESCRIPTION
**Pull request by nutanv1contr**
#### Summary:
Restrict users from uploading multiple attachments with same file name.
#### Changes:
If user tries to upload multiple attachments with same file name , then during upload it will restrict the user from uploading  with warning message "Attachment with <i> filename</i> already exist"
#### Steps to test:
1.Try to upload an attachment with same file name and extension , listed on the attachment list.
#### Expected result: 
It should give an warning message "Attachment with  <i> filename</i> already exist"

closes #417

### CODE REVIEW

**Comment by imaykay** [comment]
I am not sure, if the basic idea on how to solve this issue is the best. I have two issues with this approach:
1. I am able to upload two or more attachments with the same name in one edit cycle because the javascript only compares new names with already stored names (all attachments added in this cycle are not yet stored and linked to the object so they are not part of the comparison).
2. If we really want to achieve a better data situation in the database, the check should definitely happen on the server side. Best approach would be in the thrift services so that other clients like REST would also get this restriction.
For a fail-fast feedback, the javascript is still a good thing, but I think it is not sufficient for solving the problem.

**Comment by nutanv1contr** [comment]
I have committed the latest changes. @imaykay please review.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 1]
```diff
@@ -609,42 +619,53 @@ public RequestStatus updateRelease(Release release, User user, Iterable<Release.
...
         return RequestStatus.SUCCESS;
+       }
     }
 
-    private void deleteAttachmentUsagesOfUnlinkedReleases(Release updated, Release actual) throws SW360Exception {
+    private boolean duplicateAttachmentExist(Release release) {
+    	if(release.attachments != null && !release.attachments.isEmpty()) {
+            return RepositoryUtils.isDuplicateAttachment(release.attachments);
+        }
+		return false;
```
Comment (imaykay) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: Please check spaces vs. tabs.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java [hunk 2]
```diff
@@ -208,7 +210,14 @@ public RequestStatus updateProject(Project project, User user) throws SW360Excep
         }
     }
 
-    private void deleteAttachmentUsagesOfUnlinkedReleases(Project updated, Project actual) throws SW360Exception {
+    private boolean duplicateAttachmentExist(Project project) {
+        if(project.attachments != null && !project.attachments.isEmpty()) {
+            return RepositoryUtils.isDuplicateAttachment(project.attachments);
+        }
+        return false;
+	}
```
Comment (imaykay) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java]: Please check spaces vs. tabs.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/Sw360Portlet.java [hunk 3]
```diff
@@ -274,6 +274,9 @@ public void setSessionMessage(PortletRequest request, RequestStatus requestStatu
             case FAILED_SANITY_CHECK:
                 setSW360SessionError(request, ErrorMessages.UPDATE_FAILED_SANITY_CHECK);
                 break;
+            case DUPLICATE_ATTACHMENT:
+            	setSW360SessionError(request, ErrorMessages.DUPLICATE_ATTACHMENT);
```
Comment (imaykay) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/Sw360Portlet.java]: Please check spaces vs. tabs.

File: frontend/sw360-portlet/src/main/webapp/html/utils/includes/editAttachments.jspf [hunk 4]
```diff
@@ -168,8 +168,9 @@
         // register event handlers
         $('#upload-attachment-button').on('click', function() {
             attachmentUploader.showUploadDialog(function(addedAttachment) {
-                attachmentTable.row.add(addedAttachment).draw(false);
-                $(attachmentTable.row('#' + addedAttachment.attachmentContentId).node()).addClass('added');
+            	attachmentTable.row.add(addedAttachment).draw(false);
+            	$(attachmentTable.row('#' + addedAttachment.attachmentContentId).node()).addClass('added');
+
```
Comment (imaykay) [inline frontend/sw360-portlet/src/main/webapp/html/utils/includes/editAttachments.jspf]: Maybe you could reset this file as nothing valuable got changed?

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java [hunk 5]
```diff
@@ -23,6 +23,7 @@
     public static final String COMPONENT_DUPLICATE ="Component could not be added, since a component with the same name already exists.";
     public static final String RELEASE_NOT_ADDED = "Release could not be added.";
     public static final String RELEASE_DUPLICATE ="Release could not be added, since a release with the same name and version already exists.";
+    public static final String DUPLICATE_ATTACHMENT ="Multiple Attachments with same name cannot be present in attachment list.";
```
Comment (imaykay) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java]: I am a fan of proper formatting as you might have noticed :-) Maybe add a space after the equal sign (I know, the others did not do it consistently, but we can do it better :D ) and write "attachments" with a small A.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/RepositoryUtils.java [hunk 6]
```diff
@@ -59,4 +61,8 @@ public static RequestSummary doBulk(Collection<?> objects, User user, DatabaseRe
 
         return searchIds;
     }
+
+    public static boolean isDuplicateAttachment(Set<Attachment> attachments) {
```
Comment (imaykay) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/RepositoryUtils.java]: Maybe place this method into the AttachmentConnector, which seems to be used for attachment related methods? As this method has no functionality that can be used for different data types, I am not sure if this universal class is a good place.

**Comment by mcjaeger** [comment]
should be in for `sw360-3.4.0-M1`

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java (hunk 1 above)
Comment (nutanv1contr) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: Thanks @imaykay  for reviewing the changes.  I have added logic to check the sha1 aloong with file name check. 
As discussed about the second point we will restrict users from updating the object if there is a duplicate attachment present on the list and application will display an error message which will indicate the reason for  update failure. So I am not making any change for this as the logic is already added.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java (hunk 2 above)
Comment (nutanv1contr) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectDatabaseHandler.java]: Done.

File: frontend/sw360-portlet/src/main/webapp/html/utils/includes/editAttachments.jspf (hunk 4 above)
Comment (nutanv1contr) [reply frontend/sw360-portlet/src/main/webapp/html/utils/includes/editAttachments.jspf]: Done

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java (hunk 5 above)
Comment (nutanv1contr) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java]: Done also updated the error message as we will be checking file name along with sha1.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/RepositoryUtils.java (hunk 6 above)
Comment (nutanv1contr) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/RepositoryUtils.java]: Done.Thanks for mentioning that.

**Comment by imaykay** [comment]
I couldn't test the final changes, just looked at the code, because my data already has the defaultvendor field which causes issues in older branches :-/ @mcjaeger could you do a quick retest?

**Comment by mcjaeger** [comment]
@imaykay yes

**Comment by nutanv1contr** [comment]
Thanks @mcjaeger for your reviewing the changes. I have made necessary changes. Now application will stay in edit mode along with the error message.

**Comment by mcjaeger** [comment]
for my review, I think it would have been better to stay at the attachments tab and not switch to the general details tab in case of the error message. It is apparent at project and release attachment, for component attachments, it is OK.

**Comment by mcjaeger** [comment]
sorry, was mistakenly putting my review e-mail into the commit message, was the one of Andi of course ...

**Pull request merged**